loadgen/generic_executor.go

```diff
@@ -68,34 +71,52 @@ func (g *GenericExecutor) newRun(info ScenarioInfo) (*genericRun, error) {
 // iterations is reached.
 func (g *genericRun) Run(ctx context.Context) error {
 	ctx, cancel := context.WithCancel(ctx)
-	if g.config.Duration > 0 {
-		ctx, cancel = context.WithTimeout(ctx, g.config.Duration)
+	if g.config.Timeout > 0 {
+		g.logger.Debugf("Will timeout after %v", g.config.Timeout)
+		ctx, cancel = context.WithTimeout(ctx, g.config.Timeout)
 	}
 	defer cancel()
 
+	durationCtx := ctx
+	if g.config.Duration > 0 {
+		durationCtx, cancel = context.WithTimeout(ctx, g.config.Duration)
+		defer cancel()
+	}
+
 	startTime := time.Now()
 	var runErr error
 	doneCh := make(chan error)
 	var currentlyRunning int
-	waitOne := func() {
+	waitOne := func(exitOnDurationDone bool) {
+		timeoutOrPending := make(<-chan struct{})
+		if exitOnDurationDone {
+			timeoutOrPending = durationCtx.Done()
+		}
 		select {
 		case err := <-doneCh:
 			currentlyRunning--
 			if err != nil {
 				runErr = err
 			}
 		case <-ctx.Done():
+		case <-timeoutOrPending:
 		}
 	}
 
-	// Run all until we've gotten an error or reached iteration limit
-	for i := 0; runErr == nil && ctx.Err() == nil &&
+	if g.config.Iterations > 0 {
+		g.logger.Debugf("Will run %v iterations", g.config.Iterations)
+	} else {
+		g.logger.Debugf("Will start iterations for %v", g.config.Duration)
+	}
+
+	// Run all until we've gotten an error or reached iteration limit or timeout
+	for i := 0; runErr == nil && durationCtx.Err() == nil &&
```

Comment:
> or timeout

If the regular context completes (now backed by timeout), this loop will not exit as may be expected. Do you want to set `runErr` on `<-ctx.Done()` above or add `&& ctx.Err() == nil` here?

Reply:
Hmm, yeah, probably should. The inner stuff it's starting would time out right away but good to have it there too.